src/Control/Distributed/Process.hs

```diff
@@ -404,12 +409,22 @@ spawnChannelLocal proc = do
 callLocal ::
      Process a  -- ^ Process to run
   -> Process a  -- ^ Value returned
-callLocal proc = do
-  mv <- liftIO newEmptyMVar
-  self <- getSelfPid
-  _ <- spawnLocal $ do
-     link self
-     try proc >>= liftIO . putMVar mv
-  liftIO $ takeMVar mv >>=
-     either (throwIO :: SomeException -> IO a) return
-
+callLocal proc = mask $ \release -> do
+  mv    <- liftIO newEmptyMVar :: Process (MVar (Either SomeException a))
+  child <- spawnLocal $ do
+             ep <- try $ release $ proc
+             liftIO $ putMVar mv ep
+  liftIO (fetchResult mv) `onException`
+    (do exit child "exception in parent process"
+        waitForExit child)
+  where
+    waitForExit child = do
+      bracket (monitor child)
+              (unmonitor)
+              (\mRef -> receiveWait
```

Comment:
I think that interrupt on exception in `receiveWait` will have correct semantics here, this means that we want thread to exit immediately we could just send another exception. However, I'm not insist on that.